src/util/sha256_openssl.c

```diff
@@ -11,38 +11,39 @@
 #include "sha256.h"
 #include "util.h"
 
-#include <openssl/sha.h>
+#include <openssl/evp.h>
 
 struct SHA256State {
-	SHA256_CTX context;
+	EVP_MD_CTX *context;
 };
 
 SHA256State *sha256_new(void) {
 	SHA256State *st = calloc(1, sizeof(*st));
-	SHA256_Init(&st->context);
+	st->context = EVP_MD_CTX_new();
+	EVP_DigestInit(st->context, EVP_sha256());
 	return st;
 }
 
-void sha256_free(SHA256State *ctx) {
-	free(ctx);
+void sha256_free(SHA256State *st) {
+	EVP_MD_CTX_free(st->context);
+	free(st);
 }
 
-void sha256_update(SHA256State *ctx, const uint8_t *data, size_t len) {
-    SHA256_Update(&ctx->context, data, len);
+void sha256_update(SHA256State *st, const uint8_t *data, size_t len) {
+	EVP_DigestUpdate(st->context, data, len);
 }
 
-void sha256_final(SHA256State *ctx, uint8_t hash[SHA256_BLOCK_SIZE], size_t hashlen) {
+void sha256_final(SHA256State *st, uint8_t hash[SHA256_BLOCK_SIZE], size_t hashlen) {
 	assert(hashlen >= SHA256_BLOCK_SIZE);
-	SHA256_Final(hash, &ctx->context);
+	EVP_DigestFinalXOF(st->context, hash, hashlen);
 }
 
 void sha256_digest(const uint8_t *data, size_t len, uint8_t hash[SHA256_BLOCK_SIZE], size_t hashlen) {
 	assert(hashlen >= SHA256_BLOCK_SIZE);
 
-	SHA256_CTX ctx;
-	SHA256_Init(&ctx);
-	SHA256_Update(&ctx, data, len);
-	SHA256_Final(hash, &ctx);
+	SHA256State *st = sha256_new();
+	sha256_update(st, data, len);
+	sha256_final(st, hash, hashlen);
```

Comment:
This rewrite incurs an extra heap allocation for the `SHA256State`, which is why the old version was open-coded. This allocation is also leaked, because `sha256_free` is never called.